File: swift/Sources/IdkollenClient/Document.swift
import Foundation

public struct DocumentUploadResponse: Decodable {
    public let id: String
    public let hash: String
}

public final class DocumentEndpoint {
    private let transport: Transport
    internal init(_ transport: Transport) { self.transport = transport }

    public func upload(data: Data, filename: String, mimeType: String = "application/pdf") async throws -> DocumentUploadResponse {
        try await transport.postMultipart("/document", data: data, filename: filename, mimeType: mimeType)
    }

    public func download(id: String) async throws -> Data {
        try await transport.getRaw("/document/\(id)")
    }

    public func delete(id: String) async throws {
        try await transport.delete("/document/\(id)")
    }
}

File: dart/lib/src/document.dart
import 'transport.dart';

class DocumentUploadResponse {
  const DocumentUploadResponse({required this.id, required this.hash});
  final String id;
  final String hash;
  factory DocumentUploadResponse.fromJson(Map<String, dynamic> json) =>
      DocumentUploadResponse(id: json['id'] as String, hash: json['hash'] as String);
}

class DocumentEndpoint {
  DocumentEndpoint(this._transport);
  final Transport _transport;

  Future<DocumentUploadResponse> upload(List<int> data, String filename, {String mimeType = 'application/pdf'}) async {
    final body = await _transport.postMultipart('/document', data, filename, mimeType);
    return DocumentUploadResponse.fromJson(body);
  }

  Future<List<int>> download(String id) => _transport.getRaw('/document/$id');

  Future<void> delete(String id) => _transport.delete('/document/$id');
}
